game/songs.cc

```diff
@@ -254,63 +237,52 @@ void Songs::CacheSonglist() {
 		songObject["danceTracks"] = song->hasDance();
 		songObject["guitarTracks"] = song->hasGuitars();
 		songObject["loadStatus"] = static_cast<int>(song->loadStatus);
+
+		// Collate info
+		songObject["collateByTitle"] = song->collateByTitle;
+		songObject["collateByTitleOnly"] = song->collateByTitleOnly;
+		songObject["collateByArtist"] = song->collateByArtist;
+		songObject["collateByArtistOnly"] = song->collateByArtistOnly;
+
 		if(songObject != nlohmann::json::object()) {
-			jsonRoot.push_back(songObject);
+			jsonRoot.push_back(std::move(songObject));
 		}
 	}
 
 	fs::path cacheDir = getCacheDir() / SONGS_CACHE_JSON_FILE;
 	writeJSON(jsonRoot, cacheDir);
 	}
 
-void Songs::reload_internal(fs::path const& parent) {
-	if (std::distance(parent.begin(), parent.end()) > 20) { std::clog << "songs/info: >>> Not scanning: " << parent.string() << " (maximum depth reached, possibly due to cyclic symlinks)\n"; return; }
+void Songs::reload_internal(fs::path const& parent, Cache cache) {
 	try {
 		std::regex expression(R"((\.txt|^song\.ini|^notes\.xml|\.sm)$)", std::regex_constants::icase);
-		for (const auto &dir : fs::directory_iterator(parent)) { //loop through files
+		auto iterator = fs::recursive_directory_iterator(parent, fs::directory_options::follow_directory_symlink);
+		auto maxDepth = iterator.depth() + 10;
+		for (const auto &dir : iterator) { //loop through files
 			if (!m_loading) return; // early return in case scanning is long and user wants to exit quickly
+			if (dir.is_directory()) {
+				continue;
+			}
+			if (iterator.depth() > maxDepth) {
+				std::clog << "songs/info: >>> Not scanning: " << parent.string() << " (maximum depth reached, possibly due to cyclic symlinks)\n";
+				continue; 
+			}
 			fs::path p = dir.path();
-			if (fs::is_directory(p)) { reload_internal(p); continue; } //if the file is a folder redo this function with this folder as path
-			if (!regex_search(p.filename().string(), expression)) continue; //if the folder does not contain any of the requested files, ignore it
+			if (!regex_search(p.filename().string(), expression)) {
+				continue; //if the folder does not contain any of the requested files, ignore it
+			}
 			try { //found song file, make a new song with it.
-				{
-					std::shared_lock<std::shared_mutex> l(m_mutex);
```

Comment:
This one isn't needed because the song list isn't accessed anymore in that code block. I've restored the unique_lock in the next code block because the song list is modified then.